### ISSUE TITLE
[feature request] Expanding gradient function with variances

### ISSUE THREAD

**Issue by alexdepremia** [pytorch/pytorch#8897]
Hi,
in many applications will be useful to have available not only the gradients averaged on minibatches, but also its second momentum for each parameter.
I think developers can do it with a small computation overhead by modifying accumulate_grad.cpp (I was not able to understand the underlying class structure), accumulating also the square of the individual contribution to the gradient (which is much more simple that getting the individual contribution to the gradients as requested in https://github.com/pytorch/pytorch/issues/7786). 
I'm aware that it will duplicate the memory requirements, but hopefully it can be done as optional.
Thank you very much.

If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description

Provide a short description.

## Code example

Please try to provide a minimal example to repro the bug.
Error messages and stack traces are also helpful.

## System Info
Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

- PyTorch or Caffe2:
- How you installed PyTorch (conda, pip, source):
- Build command you used (if compiling from source):
- OS:
- PyTorch version:
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- GCC version (if compiling from source):
- CMake version:
- Versions of any other relevant libraries:

**Comment by soumith** [comment]
hi Alex.

This is not simple as modifying accumulate_grad.cpp. A lot of our functions (in-fact 99% of our neural network layers) directly compute accumuated gradients for a full mini-batch. So, for a mini-batch, we never actually get separate gradients per sample.
To introduce the feature you propose, we probably have to change ~5k+ lines of code across the ~100 odd neural-net primitives we provide, and request our vendor libraries such as CuDNN provide non-accumulated gradient functions.

This feature request came up before (cant find the issue#).
Considering the niche nature (and the 6 months to 1 year effort to pull this off), we wont provide it.
